@xen-orchestra/xapi/vdi.js

```
@@ -60,19 +62,49 @@ class Vdi {
     })
   }
 
-  async exportContent(ref, { baseRef, cancelToken = CancelToken.none, format }) {
+  async exportContent(ref, { baseRef, cancelToken = CancelToken.none, format, preferNbd = true }) {
     const query = {
       format,
       vdi: ref,
     }
     if (baseRef !== undefined) {
       query.base = baseRef
     }
+    let nbdClient
     try {
-      return await this.getResource(cancelToken, '/export_raw_vdi/', {
+      if (preferNbd) {
+        const nbdInfos = await this.call('VDI.get_nbd_info', ref)
+        if (nbdInfos.length > 0) {
+          // a little bit of randomization to spread the load
+          const nbdInfo = nbdInfos[Math.floor(Math.random() * nbdInfos.length)]
+          try {
+            nbdClient = new NbdClient(nbdInfo)
+            await nbdClient.connect()
+          } catch (err) {
+            console.error({
+              err,
+              nbdInfo,
+            })
+            nbdClient = undefined
+          }
+        }
+      }
+      // the raw nbd export does not need to peek ath the vhd source
+      if (nbdClient !== undefined && format === 'raw') {
+        return createNbdRawStream(nbdClient)
```

should throw an error if baseref is defined